x/ibc-transfer/keeper/relay.go

```diff
@@ -148,104 +127,106 @@ func (k Keeper) createOutgoingPacket(
 	return k.channelKeeper.SendPacket(ctx, channelCap, packet)
 }
 
+// OnRecvPacket processes a cross chain fungible token transfer. If the
+// sender chain is the source of minted tokens then vouchers will be minted
+// and sent to the receiving address. Otherwise if the sender chain is sending
+// back tokens this chain originally transferred to it, the tokens are
+// unescrowed and sent to the receiving address.
 func (k Keeper) OnRecvPacket(ctx sdk.Context, packet channeltypes.Packet, data types.FungibleTokenPacketData) error {
 	// NOTE: packet data type already checked in handler.go
 
-	if len(data.Amount) != 1 {
-		return sdkerrors.Wrapf(types.ErrOnlyOneDenomAllowed, "%d denoms included", len(data.Amount))
-	}
-
-	prefix := types.GetDenomPrefix(packet.GetDestPort(), packet.GetDestChannel())
-	source := strings.HasPrefix(data.Amount[0].Denom, prefix)
-
 	// decode the receiver address
 	receiver, err := sdk.AccAddressFromBech32(data.Receiver)
 	if err != nil {
 		return err
 	}
 
-	if source {
+	// This is the prefix that would have been prefixed to the denomination
+	// on sender chain IF and only if the token originally came from the
+	// receiving chain.
+	//
+	// NOTE: We use SourcePort and SourceChannel here, because the counterparty
+	// chain would have prefixed with DestPort and DestChannel when originally
+	// receiving this coin as seen in the "sender chain is the source" condition.
+	voucherPrefix := types.GetDenomPrefix(packet.GetSourcePort(), packet.GetSourceChannel())
```

Comment:
Once you change `ReceiverChainIsSource`, I would move this line into the if clause. Since this is a prefix that really is only relevant in the `ReceiverChainIsSource` case, it makes sense to define it in the if clause and not at the top-level of this function.

It also makes it easier to understand what each clause is doing since the code will be structured the exact same way in each clause. They both will:

1. Determine the prefix
2. Do Prefixing logic (either add or remove prefix)
3. Send Coins to address (either by minting or unescrowing)

The consistency in structure both for the `SenderIsSource` and `ReceiverIsSource` AND in the 2 cases for receiving will make it much easier for new developers to understand what the code is doing and why